lib/msf/core/exploit/smb/client/psexec.rb

```diff
@@ -193,6 +196,246 @@ def psexec(command, disconnect=true)
     true
   end
 
-end
+  def powershell_installed?(smb_share, psh_path)
+    share = "\\\\#{datastore['RHOST']}\\#{smb_share}"
+
+    case smb_share.upcase
+    when 'ADMIN$'
+      path = 'System32\\WindowsPowerShell\\v1.0\\powershell.exe'
+    when 'C$'
+      path = 'Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe'
+    else
+      path = psh_path
+    end
+
+    simple.connect(share)
+    vprint_status("Checking for #{path}")
+    if smb_file_exist?(path)
+      vprint_status('PowerShell found')
+      psh = true
+    else
+      vprint_status('PowerShell not found')
+      psh = false
+    end
+
+    simple.disconnect(share)
+    psh
+  end
+
+  def execute_command(text, bat, cmd)
+    # Try and execute the provided command
+    execute = "%COMSPEC% /C echo #{cmd} ^> %SYSTEMDRIVE%#{text} > #{bat} & %COMSPEC% /C start %COMSPEC% /C #{bat}"
+    vprint_status("Executing the command...")
+    begin
+      return psexec(execute)
+    rescue Rex::Proto::DCERPC::Exceptions::Error, Rex::Proto::SMB::Exceptions::Error => e
+      elog("#{e.class} #{e.message}\n#{e.backtrace * "\n"}", 'rex', LEV_3)
+      print_error("Unable to execute specified command: #{e}")
+      return false
+    end
+  end
+
+  def execute_command_with_output(text, bat, cmd, smb_share, r_ip, delay, rt)
+
+      res = execute_command(text, bat)
+
+      if res
+        for i in 0..(rt)
+          Rex.sleep(delay)
+          # if the output file is still locked then the program is still likely running
+          if (exclusive_access(text))
+            break
+          elsif (i == rt)
+            print_error("Command seems to still be executing. Try increasing RETRY and DELAY")
+          end
+        end
+        get_output(text, smb_share, r_ip)
+      end
+
+      cleanup_after(bat, smb_share, r_ip)
+    end
+
+  # Retrive output from command
+  def get_output(file, smb_share, r_ip)
+    print_status("Getting the command output...")
+    output = smb_read_file(smb_share, r_ip, file)
+    if output.nil?
+      print_error("Error getting command output. #{$!.class}. #{$!}.")
+      return
+    end
+    if output.empty?
+      print_status("Command finished with no output")
+      return
+    end
+
+  end
+
+  # check if our process is done using these files
+  def exclusive_access(*files, smb_share, r_ip)
```

Comment:
`exclusive_access`, `get_output` and `cleanup_after` should be made private now.  Either by moving them to the end of the file after a `private` keyword or by adding the restriction inside the class scope around L439
```
private :exclusive_access
private :get_output
private :cleanup_after
```

See https://en.wikibooks.org/wiki/Ruby_Programming/Syntax/Classes#Private